client/core/simnet_trade.go

```diff
@@ -1151,35 +1211,18 @@ func (s *simulationTest) monitorOrderMatchingAndTradeNeg(ctx context.Context, cl
 	}
 	tracker.mtx.RUnlock()
 
-	return s.monitorTrackedTrade(client, tracker, finalStatus)
-}
-
-func (s *simulationTest) runMiners() context.CancelFunc {
-	ctx, cancel := context.WithCancel(context.Background())
-
-	runMiner := func(a *assetConfig, swapConfs uint32) {
-		ctrl := newHarnessCtrl(a.id)
-		go func() {
-			for {
-				ctrl.mineBlocks(s.ctx, swapConfs)
-				// wait between 1 and 5 seconds.
-				delay := time.Duration(rand.Float64()*float64(time.Second)*4) + time.Second
-				select {
-				case <-time.After(delay):
-				case <-ctx.Done():
-					return
-				}
-			}
-		}()
+	if err := s.calcBalanceChanges(client, orderID, finalStatus); err != nil {
+		return err
 	}
 
-	runMiner(s.base, 1)
-	runMiner(s.quote, 1)
-
-	return cancel
+	return s.monitorTrackedTrade(client, tracker, finalStatus)
 }
 
-func (s *simulationTest) monitorTrackedTrade(client *simulationClient, tracker *trackedTrade, finalStatus order.MatchStatus) error {
+func (s *simulationTest) calcBalanceChanges(client *simulationClient, orderID string, finalStatus order.MatchStatus) error {
```

Comment:
`calcBalanceChanges` is only used from within `monitorOrderMatchingAndTradeNeg`, where we've already done `findOrder`, and do a `range tracker.matches` loop with proper `tracker.mtx` use.

I think `monitorOrderMatchingAndTradeNeg` should look like:

```go
func (s *simulationTest) monitorOrderMatchingAndTradeNeg(ctx context.Context, client *simulationClient, orderID string, finalStatus order.MatchStatus) error {
	errs := newErrorSet("[client %s] ", client.name)

	tracker, err := client.findOrder(orderID)
	if err != nil {
		return errs.addErr(err)
	}

	// Wait up to 2 times the epoch duration for this order to get matched.
	maxMatchDuration := 2 * time.Duration(tracker.epochLen) * time.Millisecond
	client.log.Infof("Waiting up to %v for matches on order %s", maxMatchDuration, tracker.token())
	matched := client.notes.find(ctx, maxMatchDuration, func(n Notification) bool {
		orderNote, isOrderNote := n.(*OrderNote)
		isMatchedTopic := n.Topic() == TopicBuyMatchesMade || n.Topic() == TopicSellMatchesMade
		return isOrderNote && isMatchedTopic && orderNote.Order.ID.String() == orderID
	})
	if ctx.Err() != nil { // context canceled
		return nil
	}
	if !matched {
		return errs.add("order %s not matched after %s", tracker.token(), maxMatchDuration)
	}

	recordBalanceChanges := func(isSwap bool, qty, rate uint64) {
		amt := int64(qty)
		a := s.base
		if client.isSeller != isSwap {
			// use quote amt for seller redeem and buyer swap
			amt = int64(calc.BaseToQuote(rate, qty))
			a = s.quote
		}
		if isSwap {
			amt *= -1
		}
		client.log.Infof("updated %s balance diff with %s", a.symbol, a.valFmt(amt))
		client.expectBalanceDiffs[a.id] += amt
	}

	tracker.mtx.RLock()
	client.log.Infof("%d match(es) received for order %s", len(tracker.matches), tracker.token())
	for _, match := range tracker.matches {
		client.log.Infof("%s on match %s, amount %s %s", match.Side.String(), token(match.MatchID.Bytes()),
			s.base.valFmt(match.Quantity), s.base.symbol)
		if match.Side == order.Taker {
			if finalStatus >= order.TakerSwapCast {
				recordBalanceChanges(true, match.Quantity, match.Rate)
			}
			if finalStatus >= order.MatchComplete {
				recordBalanceChanges(false, match.Quantity, match.Rate)
			}
		}
		if match.Side == order.Maker {
			if finalStatus >= order.MakerSwapCast {
				recordBalanceChanges(true, match.Quantity, match.Rate)
			}
			if finalStatus >= order.MakerRedeemed {
				recordBalanceChanges(false, match.Quantity, match.Rate)
			}
		}
	}
	tracker.mtx.RUnlock()

	return s.monitorTrackedTrade(client, tracker, finalStatus)
}
```